src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs

```diff
@@ -190,20 +191,18 @@ private DotnetHost DotnetHostInfo
                     AddCommandLineCommands(commandLineBuilder);
                     var commandLine = commandLineBuilder.ToString();
 
-                    if (string.IsNullOrEmpty(ToolPath))
+                    // ToolExe delegates back to ToolName if the override is not
+                    // set.  So, if ToolExe != ToolName, we know ToolExe is
+                    // explicitly overriden - so use it as a native invocation.
+                    if (string.IsNullOrEmpty(ToolPath) || ToolExe == ToolName)
                     {
-                        _dotnetHostInfo = DotnetHost.CreateManagedToolInvocation(ToolNameWithoutExtension, commandLine);
-
-                        // See comment in ManagedCompiler.cs on why this if statement is here.
-                        if (ToolExe != _dotnetHostInfo.ToolNameOpt)
-                        {
-                            _dotnetHostInfo = DotnetHost.CreateUnmanagedToolInvocation(ToolPath, commandLine);
-                        }
+                        _dotnetHostInfo = DotnetHost.CreateManagedToolInvocation(ToolName, commandLine);
                     }
                     else
                     {
-                        // Explicitly provided ToolPath, don't try to figure anything out
-                        _dotnetHostInfo = DotnetHost.CreateUnmanagedToolInvocation(ToolPath, commandLine);
+                        // Explicitly provided ToolPath or ToolExe, don't try to
+                        // figure anything out
+                        _dotnetHostInfo = DotnetHost.CreateNativeToolInvocation(Path.Combine(ToolPath, ToolExe), commandLine);
```

Comment:
Still dont get why we are assuming that this is native tool. All we know is the name. It is actually common practice to use CscToolPath to run custom versions of roslyn this way.

Reply:
Please give me a better name then :)